meteor-bundle-main.js

```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
+        if (now > metadata.notAfter) {
+          console.log("Skipping", keyFilename, "because", metadata.notAfter, "is in the past.");
           continue;
         }
 
-        result['ca'] = metadata['ca'];
-        result['key'] = fs.readFileSync(keyFilename, 'utf-8');
-        result['cert'] = metadata['cert'];
-        return result;
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
+        console.error("Aieee, we found zero valid certificates. Allowing Sandstorm to crash.");
+        return;
       }
+
+      console.log("Using this HTTPS key:", validCertificates[0].keyFilename,
+                  "valid starting", new Date(validCertificates[0].notBefore),
+                  "until", new Date(validCertificates[0].notAfter));
+
+      // Store HTTPS configuration data from the oldest certificate.
+      var result = {
+        ca: validCertificates[0].ca,
+        key: fs.readFileSync(validCertificates[0].keyFilename, 'utf-8'),
+        cert: validCertificates[0].cert
+      };
+
+      // Calculate re-key time.
+      //
+      // - If the cert we want to switch to is not valid yet, then we
+      //   re-key at its notBefore time.
+      //
+      // - If the cert we want to switch to *is* already valid, then
+      //   we re-key at the notAfter time of our current cert.
+      //
+      // - If there is no cert we want to switch to, then the
+      //   nextRekeyTime is null.
+      result.nextRekeyTime = null;
+
+      if (validCertificates.length >= 2) {
+        var secondBest = validCertificates[1];
+        if (now < secondBest.notBefore) {
+          result.nextRekeyTime = secondBest.notBefore;
+        } else {
+          result.nextRekeyTime = validCertificates[0].notAfter;
+        }
+
+        console.log("Will switch to", validCertificates[1].keyFilename,
+                    "at time", new Date(result.nextRekeyTime),
+                    "via real-time SNI re-keying system.");
+      } else {
+        console.log("We have only the one key that we know about. We'll need to renew soon.");
+      }
+
+      return result;
     };
 
     if (process.env.HTTPS_PORT) {
-      var httpsServer = https.createServer(getHttpsOptions(), requestListener);
+      // Great! FD #3 will speak HTTPS.
+      //
+      // NOTE: This assumes that the user will only set HTTPS_PORT if
+      // there are valid certificates for us to use. This could be a
+      // problem if BASE_URL is https but we aren't ready.
+
+      // Create a local variable with key information.
+      var sandcatsState = {};
+
+      // Create a global that others can call into if they want to
+      // re-key, such as when `sandcats.js` downloads a new
+      // certificate.
+      global.sandcats = {};
```

Comment:
Ya, I'm using a global here because I'm at the most top-level file, before Meteor exists yet, and I want to provide access to some state in this file. To try to be clean, I do it by proving a function that encapsulates the important aspect of state, rather than providing the raw state.

Other ideas welcome.